src/State/Editing.hs

```diff
@@ -190,13 +194,30 @@ handleEditingInput e = do
                         csEditState.cedEditor %= applyEdit Z.moveRight
                     | otherwise -> doInsertChar
 
-          _ | editingPermitted st -> mhHandleEventLensed (csEditState.cedEditor) handleEditorEvent e
+          _ | editingPermitted st -> do
+              mhHandleEventLensed (csEditState.cedEditor) handleEditorEvent e
+              sendUserTypingAction
```

Comment:
This might be a bigger ask than some of the others, but: sending a typing action notification on every keystroke is going to result in a lot of HTTP traffic on the websocket. My concern is that the frequency of those outgoing events might be much greater than the frequency required to show the typing notification on other clients, which means a lot of outgoing traffic would be wasted. Is this how the web client implements this? What I'm wondering is whether it would be possible to abstract this a bit so that the outgoing typing notification frequency is capped at, say, N/2, where N is the expiration period of the notification on other (typical) clients. So if other clients would only need a refresh at around N seconds to keep showing the notification, we could send these out every N/2 seconds at most to keep those client notifications fresh but avoid wasting outbound traffic when someone is typing quickly. :)

Reply:
On another note: this might not be the only editing handler that needs to trigger these events. Lots of things will indeed fall through to `handleEditorEvent`, but any place we've implemented other editing operations like word kill, etc., will also need to call `sendUserTypingAction`.

Reply:
I checked the web client code and it uses 5 seconds as the expiration period (as implemented [here](https://github.com/mattermost/mattermost-webapp/blob/c41de179cdb82435f7b32e346811f7cad20603dd/stores/user_typing_store.jsx#L47) and as set [here](https://github.com/mattermost/mattermost-server/blob/7ed1177a2b676aa4c93515268642c855cfe57a37/config/default.json#L50)).

I'll change `sendUserTypingAction` to send notification to the server once every 2 seconds.

Reply:
Thanks! What do you have in mind? I'm imagining a state field (per-channel) storing the last time a typing notification was sent, where `sendUserTypingAction` ignores requests unless 2 seconds have elapsed, in which case the timestamp gets updated and an action gets sent.

Reply:
When we move to the `TChan` and a separate thread to push websocket actions, we can just add a timestamp along with the action in the channel (that is, the channel is of type `TChan (WebsocketAction, UTCTime)`) and the pusher thread can discard the actions which occur with 2 seconds of the last action's timestamp. The thread can keep this timestamp a local variable, just like it'll keep the current action sequence number. This way, we avoid adding anything into the `ChatState`.

How does this sound?

Reply:
Well, with that approach I'm concerned that later there might be `WebsocketAction` values that have no need for this behavior, or for which delaying sending on a timestamp basis would be broken. The specific action of the user typing in a channel may end up being the only one that has this need.

How about this: rather than sending a raw websocket action over the channel, make a new type to abstract it (with a better name than this):
```
data CustomAction = UserTyping UTCTime ChannelId
```
and then:
* the metadata required for that specific action is required to be available when the action is sent,
* the thread can still do local state management to track when to send actions, and
* the thread can translate these higher-level application-specific notions into the lower-level websocket actions as appropriate.

Reply:
(It might be better, to avoid confusion, to call the above constructor `SelfTyping` or something since it is about the user running Matterhorn rather than some other user who is typing.)

Reply:
Alright, I'll do it this way.

Reply:
Thanks!